src/installer/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs

```diff
@@ -103,6 +98,41 @@ void SetLastWriteTime()
 
             try
             {
+                RetryUtil.RetryOnIOError(() =>
+                {
+                    MemoryMappedFile memoryMappedFile = null;
+                    MemoryMappedViewAccessor memoryMappedViewAccessor = null;
+                    try
+                    {
+                        // Open the source host file.
+                        memoryMappedFile = MemoryMappedFile.CreateFromFile(appHostSourceFilePath, FileMode.Open, null, 0, MemoryMappedFileAccess.CopyOnWrite);
+                        memoryMappedViewAccessor = memoryMappedFile.CreateViewAccessor(0, 0, MemoryMappedFileAccess.CopyOnWrite);
+
+                        // Transform the host file in-memory.
+                        RewriteAppHost(memoryMappedViewAccessor);
+
+                        // Save the transformed host.
+                        BinaryUtils.SaveFile(memoryMappedViewAccessor, appHostDestinationFilePath);
+                    }
+                    finally
+                    {
+                        if (memoryMappedViewAccessor != null)
+                        {
+                            memoryMappedViewAccessor.Dispose();
+                        }
+                        if (memoryMappedFile != null)
+                        {
+                            memoryMappedFile.Dispose();
+                        }
+                    }
+                });
+
+                RetryUtil.RetryOnWin32Error(UpdateResources);
+
+                RetryUtil.RetryOnIOError(RemoveSignatureIfMachO);
```

Comment:
Won't both of these reopen the file?  Can they be rewritten to use the memory mapped file?

Reply:
Yes, both of these will re-open the file.  `UpdateResources` uses win32 APIs to enumerate and add the resources.  There might be a way to do this against the existing open `MemoryMappedFile`, but likely the cost of doing so is going to be higher than we want to pay, if it's possible.  It would require that we re-write the functionality.

Reply:
There's a trick where you can keep an open handle read/write with sharing and it will prevent some other process from getting exclusive access.  So even if you didn't re-implement these over the file mapping you might be able to keep a handle open to prevent interference.  This works so long as the Windows APIs you're calling specify read/write sharing with the held handle.  If the Windows APIs themselves don't allow sharing your held handle prevents them from opening the file.

Reply:
I did some experiments with this, but it doesn't look like this is possible.  The file is loaded via `LoadLibraryEx`, which does not appear to allow for sharing when an existing handle has write permissions.

Reply:
I see, I guess when UpdateResource is rewritten that would fix this.  Presumably some code can be copied from CSC since it knows how to do that.

For `RemoveSignatureIfMachO` just make your new `BinaryUtils.SaveFile` take a stream and open the stream above.  Then change `RemoveSignatureIfMachO` to take a stream and give it the written file stream.  This is minimal change and saves one roll-of-the-dice around retries.

Reply:
Good call.  At this point it is for sure safe because resources can only be copied into a PEFile, and the signature can only be removed from MachO binaries.  So the operations cannot be run on the same binary.

Reply:
That's an interesting point.  I have filed https://github.com/dotnet/sdk/issues/16173 to track this.